tests/test_initialization.py

```diff
@@ -1544,3 +1550,112 @@ def new_state_dict():
+            )
+            sd = get_eva_state_dict(model, peft_config, dataloader)
+            state_dicts.append(sd["transformer.h.0.attn.c_attn"])
+
+        cos_sims = []
+        for i, j in itertools.combinations(range(self.NUM_SEEDS), 2):
+            cos_sims.append(torch.cosine_similarity(state_dicts[i], state_dicts[j], dim=1).abs())
+
+        mean_cosine_similarity = torch.cat(cos_sims).mean()
+        assert mean_cosine_similarity > self.COSINE_SIMILARITY_THRESHOLD, (
+            f"Mean absolute cosine similarity {mean_cosine_similarity:.4f} "
+            f"is not greater than {self.COSINE_SIMILARITY_THRESHOLD}"
+        )
+
+    def test_eva_initialization_consistency(self, peft_model, dataset, peft_config):
+        state_dicts = []
+        for seed in range(3):
+            shuffled_dataset = dataset.shuffle(seed=seed)
+            dataloader = DataLoader(
+                shuffled_dataset,
+                batch_size=self.BATCH_SIZE,
+                collate_fn=lambda examples: {
+                    k: torch.stack([v[k] for v in examples], dim=0) for k in examples[0].keys()
+                },
+                shuffle=False,
+            )
+            initialize_lora_eva_weights(peft_model, peft_config, dataloader)
+            state_dicts.append(peft_model.model.transformer.h[0].attn.c_attn.lora_A["default"].weight)
+
+        cos_sims = []
+        for i, j in itertools.combinations(range(self.NUM_SEEDS), 2):
+            cos_sims.append(torch.cosine_similarity(state_dicts[i], state_dicts[j], dim=1).abs())
+
+        mean_cosine_similarity = torch.cat(cos_sims).mean()
+        assert mean_cosine_similarity > self.COSINE_SIMILARITY_THRESHOLD, (
+            f"Mean absolute cosine similarity {mean_cosine_similarity:.4f} "
+            f"is not greater than {self.COSINE_SIMILARITY_THRESHOLD}"
+        )
+
+    def test_eva_config_validation(self, peft_config):
+        """Test that EvaConfig.__init__ raises a ValueError when rho is negative."""
+        with pytest.raises(ValueError, match="`rho` must be >= 1."):
+            invalid_config = deepcopy(peft_config)
+            invalid_config.eva_config = EvaConfig(rho=-1)
+
+    def test_lora_config_raises_error_without_eva_config_but_eva_init(self):
+        """Test that LoraConfig.__init__ raises a ValueError when init_lora_weights is 'eva' but eva_config is not set."""
+        with pytest.raises(ValueError, match="`eva_config` must be specified when `init_lora_weights` is 'eva'."):
+            LoraConfig(init_lora_weights="eva")
+
+    def test_lora_config_raises_warning_with_eva_config_but_not_eva_init(self):
+        """Test that LoraConfig.__init__ raises a warning when init_lora_weights is not 'eva' but eva_config is set."""
+        with pytest.warns(
+            UserWarning, match="`eva_config` specified but will be ignored when `init_lora_weights` is not 'eva'."
+        ):
+            LoraConfig(init_lora_weights=True, eva_config=EvaConfig())
```

Comment:
Can we have small tests for:

- `whiten=True`
- `prepare_activations_fn` being a mapping

Reply:
I added a whitening setting for `test_eva_state_dict_consistency`. To test `prepare_activations_fn` I added a test `test_eva_state_dict_prepare_activations_mapping`